### ISSUE TITLE
Virtual Endpoint responses aren't cached

### ISSUE THREAD

**Issue by LowCostCustoms** [TykTechnologies/tyk#3853]
**Branch/Environment/Version**
- Branch/Version: v3.2.2
- Environment: On-prem

**Describe the bug**
When caching is configured for API, and request is handled by a virtual endpoint, the response is not cached even though caching middleware uses special handling for virtual endpoints.

**Reproduction steps**

1. Add api
2. Turn on caching
3. Add virtual endpoint, which makes a request against a slow service
4. Turn on caching for the virtual endpoint
5. Make a request against the endpoint
6. Make a request against the endpoint one more time

**Actual behavior**
The second call takes as much time as the first call, the response is not cached.

**Expected behavior**
Virtual endpoint response is cached and the subsequent calls return the response from cache.

**Logs (debug mode or log file):**
Log from console or from log file.

**Configuration (tyk config file):**
```json
{
  "name": "apiname",
  "api_id": "apiid",
  "org_id": "orgid",
  "proxy": {
    "listen_path": "/",
    "target_url": "",
    "strip_listen_path": true,
    "preserve_host_header": false,
    "disable_strip_slash": true
  },
  "version_data": {
    "not_versioned": true,
    "versions": {
      "Default": {
        "name": "Default",
        "use_extended_paths": true,
        "extended_paths": {
          "virtual": [
            {
              "response_function_name": "handler",
              "function_source_type": "blob",
              "function_source_uri": "ZnVuY3Rpb24gaGFuZGxlcihyZXF1ZXN0LCBzZXNzaW9uLCBjb25maWcpIHsKICB2YXIgcmVxdWVzdE9iamVjdCA9IHsKICAgICJNZXRob2QiOiAiR0VUIiwKICAgICJEb21haW4iOiAiaHR0cHM6Ly9teS1zbG93LXdlYnNpdGUuY29tIiwKICAgICJSZXNvdXJjZSI6ICIvYXBpL3YxL2VuZHBvaW50IiwKICB9OwogIHZhciByZW1vdGVSZXNwb25zZU9iamVjdCA9IEpTT04ucGFyc2UoVHlrTWFrZUh0dHBSZXF1ZXN0KEpTT04uc3RyaW5naWZ5KHJlcXVlc3RPYmplY3QpKSk7CgogIHZhciByZXNwb25zZU9iamVjdCA9IHsKICAgICJCb2R5IjogcmVtb3RlUmVzcG9uc2VPYmplY3QuYm9keSwKICAgICJDb2RlIjogMjAwLAogICAgIkhlYWRlcnMiOiB7CiAgICAgICJDb250ZW50LVR5cGUiOiAiYXBwbGljYXRpb24vanNvbiIsCiAgICAgICJ4LXR5ay1jYWNoZS1hY3Rpb24tc2V0IjogIjEiLAogICAgICAieC10eWstY2FjaGUtYWN0aW9uLXNldC10dGwiOiAiMTIwIiwKICAgIH0sCiAgfTsKCiAgcmV0dXJuIFR5a0pzUmVzcG9uc2UocmVzcG9uc2VPYmplY3QsIHNlc3Npb24ubWV0YV9kYXRhKTsKfQo=",
              "path": "",
              "method": "GET",
              "use_session": false,
              "proxy_on_error": false
            }
          ],
          "cache": [".*"]
        }
      }
    }
  },
  "cache_options": {
    "enable_cache": true,
    "cache_all_safe_requests": true,
    "enable_upstream_cache_control": true
  },
  "enable_batch_request_support": true,
  "use_keyless": true,
  "internal": false
}
```

**Additional context**
Reordering middleware initialization order (`VirtualEndpoint` middleware is placed after `RedisCacheMiddleware`) solves the issue.

**Comment by agata-wit** [comment]
Hey @LowCostCustoms, thanks for raising this.
When you say that "reordering middleware initialization order solves the issue" - where exactly are you reordering it?

**Comment by LowCostCustoms** [comment]
Hi, @agata-wit!
The problem disappears when I move `gw.mwAppendEnabled(&chainArray, &VirtualEndpoint{BaseMiddleware: baseMid})` (file `gateway/api_loader.go:436`) right under `gw.mwAppendEnabled(&chainArray, &RedisCacheMiddleware{BaseMiddleware: baseMid, CacheStore: &cacheStore})` (file `gateway/api_loader.go:460`).

**Comment by agata-wit** [comment]
Thanks!
Just FYI we are currently working on improving endpoint designer, which means improving the middleware as well. This looks like a relatively simple fix that we could take on board as part of the work we do on endpoint designer. Let me take this up with the squad, ok?
The improved endpoint designer is planned to be released in the second half of 2022, but we might be able to include small fixes like this earlier than that. I'll let you know once we have a final decision.

**Comment by nvta-sbiyyala** [comment]
> Thanks! Just FYI we are currently working on improving endpoint designer, which means improving the middleware as well. This looks like a relatively simple fix that we could take on board as part of the work we do on endpoint designer. Let me take this up with the squad, ok? The improved endpoint designer is planned to be released in the second half of 2022, but we might be able to include small fixes like this earlier than that. I'll let you know once we have a final decision.

Thanks @agata-wit ! Along with that, if you could please pull in https://github.com/TykTechnologies/tyk/issues/3858, that would be _great_!

**Comment by agata-wit** [comment]
@nvta-sbiyyala yes, definitely. It's a bigger theme we're working on here, so anything that's currently labelled as 'middleware' and/or 'endpoint designer' is a candidate for that theme. I have my eye on #3858 as well!

**Comment by andyo-tyk** [comment]
This was fixed in Tyk 4.0.10 and Tyk 4.3.1.
